x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -67,22 +95,80 @@ public RepositoryMetaData getMetadata() {
+        String leaderIndex = index.getName();
+        Client remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
+
+        ClusterStateResponse response = remoteClient
+            .admin()
+            .cluster()
+            .prepareState()
+            .clear()
+            .setMetaData(true)
+            .setIndices(leaderIndex)
+            .get();
+
+        // Validates whether the leader cluster has been configured properly:
+        PlainActionFuture<String[]> future = PlainActionFuture.newFuture();
+        IndexMetaData leaderIndexMetaData = response.getState().metaData().index(leaderIndex);
+        ccrLicenseChecker.fetchLeaderHistoryUUIDs(remoteClient, leaderIndexMetaData, future::onFailure, future::onResponse);
+        String[] leaderHistoryUUIDs = future.actionGet();
+
+        IndexMetaData.Builder imdBuilder = IndexMetaData.builder(leaderIndexMetaData);
+        // Adding the leader index uuid for each shard as custom metadata:
+        Map<String, String> metadata = new HashMap<>();
+        metadata.put(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_SHARD_HISTORY_UUIDS, String.join(",", leaderHistoryUUIDs));
+        metadata.put(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_UUID_KEY, leaderIndexMetaData.getIndexUUID());
+        metadata.put(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_NAME_KEY, leaderIndexMetaData.getIndex().getName());
+        metadata.put(Ccr.CCR_CUSTOM_METADATA_REMOTE_CLUSTER_NAME_KEY, remoteClusterAlias);
+        imdBuilder.putCustom(Ccr.CCR_CUSTOM_METADATA_KEY, metadata);
+
+        imdBuilder.settings(leaderIndexMetaData.getSettings());
+
+        // Copy mappings from leader IMD to follow IMD
+        for (ObjectObjectCursor<String, MappingMetaData> cursor : leaderIndexMetaData.getMappings()) {
+            imdBuilder.putMapping(cursor.value);
+        }
+
+        imdBuilder.setRoutingNumShards(leaderIndexMetaData.getRoutingNumShards());
+
+        return imdBuilder.build();
     }
 
     @Override
     public RepositoryData getRepositoryData() {
-        throw new UnsupportedOperationException("Unsupported for repository of type: " + TYPE);
+        Client remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
+        ClusterStateResponse response = remoteClient.admin().cluster().prepareState().clear().setMetaData(true).get();
+        MetaData remoteMetaData = response.getState().getMetaData();
+
+        Map<String, SnapshotId> copiedSnapshotIds = new HashMap<>();
+        Map<String, SnapshotState> snapshotStates = new HashMap<>(copiedSnapshotIds.size());
+        Map<IndexId, Set<SnapshotId>> indexSnapshots = new HashMap<>(copiedSnapshotIds.size());
+
+        ImmutableOpenMap<String, IndexMetaData> remoteIndices = remoteMetaData.getIndices();
+        for (String indexName : remoteMetaData.getConcreteAllIndices()) {
+            SnapshotId snapshotId = new SnapshotId(indexName, SNAPSHOT_UUID);
```

Comment:
instead of one snapshot per index, I think it's more natural to have a fixed SnapshotID(latest, latest) and all the indices of the cluster then as indices that are part of the snapshot.

Reply:
done